### ISSUE TITLE
Provide flexible options to traverse the resrc hierarchy

### ISSUE THREAD

**Issue by lipari** [flux-framework/flux-sched#85]
This issue captures the request to design a level of abstraction in resrc that provides the same flexibility and functionality that the hwloc API provides.  This would include task mapping options.
https://github.com/flux-framework/flux-sched/pull/79#issuecomment-147103629

**Comment by dongahn** [comment]
Since we wanted to collect any many use cases as possible, here is one case I've encountered. 

Using hwloc resource data, I've created a data structure that can map a resource signature <hostname, SHA of the per-rank hwloc xml export> to a broker rank (or ranks). This logic (`res2ranks` class ) will allow sched to target the correct rank in assigning the core count, but I'd also like to use the same--well slightly extended--structure to perform a consistency check between this 'self-discovered' hwloc data and rdl-specified resrc data, when rdl/resrc is used as reader mode. 

For this check, one thing I thought would be helpful is a resrc_tree capability to iterate over resrc objects of a certain type: node-type more specifically. Then, I can simply iterate over them and perform a "equality" check against the `res2ranks` structure. To begin with, consistency check can be as simple as "given this node resrc, has hwloc discovered a machine with the same name, socket count and core count?" Of course the consistency check can be further extended if more comprehensive check is required.

Looking at `resrc`, can we consider something like `resrc_tree_t *resrc_tree_get_by_type (resrc_tree_t *topo, const char *res_type, int index)` for a use case like this? This would be similar to `hwloc_get_obj_by_type` of hwloc. 

BTW, we won't need a consistency check for hwloc reader option because in this case resrc is generated out of the hwloc data itself.
